rfcs/0010-target-registered-compiler-flow-customisation.md

```diff
@@ -0,0 +1,186 @@
+- Feature Name: `Target` registered compiler flow customisation
+- Start Date: 2021-07-14
+- RFC PR: https://github.com/apache/tvm-rfcs/pull/10
+- GitHub Issue: https://github.com/apache/tvm/issues/8589
+
+# Summary
+[summary]: #summary
+
+In order to enable flexibility in how individual targets are lowered and built within TVM, this RFC proposes additional hooks on the `Target` and that the target becomes the central place for such hooks, for example:
+
+```c++
+using FTVMRelayToTIR = Pass;
+using FTVMTIRToRuntime = runtime::TypedPackedFunc<runtime::Module(IRModule, Target)>;
+
+TVM_REGISTER_TARGET_KIND("cmsisnn", kDLCPU)
+    .set_attr<FTVMRelayToTIR>("RelayToTIR", CMSISNNLowering)
+    .set_attr<FTVMTIRToRuntime>("TIRToRuntime", CMSISNNCodeGen);
+```
+
+This defines two new hooks as attributes on the target, referencing functions registered into the central TVM registry. In similar fashion, external code generators (registered under the `relay.ext.` namespace currently) would be grouped with an appropriate `Target` as well:
+
+```c++
+using FTVMRelayToRuntime = runtime::TypedPackedFunc<runtime::Module(const Function&)>;
+using FTVMConstantUpdater = runtime::TypedPackedFunc<Map<String, runtime::NDArray>(Expr, std::string)>;
+
+TVM_REGISTER_TARGET_KIND("ethos-n", kDLCPU)
+    .set_attr<FTVMRelayToRuntime>("RelayToRuntime", EthosNCodeGen)
+    .set_attr<FTVMConstantUpdater>("UpdateConstants", EthosNConstantUpdater);
+```
+
+Collecting all targets under the `Target` functionality (as opposed to registering additional `Target`s through the function registry using the namespace `relay.ext.`) and making it clearer which hooks apply to each target.
+
+# Motivation
+[motivation]: #motivation
+
+Currently to introduce an external code generator (otherwise known as [BYOC](https://tvm.apache.org/docs/dev/relay_bring_your_own_codegen.html)), the entire compilation pipeline must be recreated; this is necessary for some targets but in the case of simply re-using existing libraries or introducing a function call to use for an operator it can become more than is necessary; to implement an external code generator requires going directly from Relay to a `runtime::Module` and re-implementing any compiler passes and code generation functionality rather than being able to extend upon the existing compiler infrastructure.
```

Comment:
nit: May be worth a pithy opening sentence saying we want to make BOYC more modular: instead of going from Relay to Module in one big step you can a) break it into phases and b) make use of existing transformations between phases. 

 Or at least that's the dumbed-down way I've been thinking of it.

Reply:
I have liberally taken that phrase and put it in :smile_cat: thanks!